// zombie/lib/zombie/cookies.js
var Cookies, URL, core, dequote, deserialize, serialize;
URL = require("url");
core = require("jsdom").dom.level3.core;
serialize = function(browser, domain, path, name, cookie) {
  var str;
  str = "" + name + "=" + cookie.value + "; domain=" + domain + "; path=" + path;
  if (cookie.expires) {
    str = str + ("; max-age=" + (cookie.expires - browser.clock));
  }
  if (cookie.secure) {
    str = str + "; secure";
  }
  return str;
};
deserialize = function(serialized) {
  var cookie, field, fields, first, key, name, val, value, _i, _len, _ref, _ref2;
  fields = serialized.split(/;+/);
  first = fields[0].trim();
  _ref = first.split(/\=/, 2), name = _ref[0], value = _ref[1];
  cookie = {
    name: name,
    value: value
  };
  for (_i = 0, _len = fields.length; _i < _len; _i++) {
    field = fields[_i];
    _ref2 = field.trim().split(/\=/, 2), key = _ref2[0], val = _ref2[1];
    switch (key.toLowerCase()) {
      case "domain":
        cookie.domain = dequote(val);
        break;
      case "path":
        cookie.path = dequote(val).replace(/%[^\/]*$/, "");
        break;
      case "expires":
        cookie.expires = new Date(dequote(val));
        break;
      case "max-age":
        cookie['max-age'] = parseInt(dequote(val), 10);
        break;
      case "secure":
        cookie.secure = true;
    }
  }
  return cookie;
};
dequote = function(value) {
  return value.replace(/^"(.*)"$/, "$1");
};
Cookies = (function() {
  function Cookies(browser, cookies, hostname, pathname) {
    var domainMatch, selected;
    if (!pathname || pathname === "") {
      pathname = "/";
    }
    domainMatch = function(domain, hostname) {
      if (domain === hostname) {
        return true;
      }
      return domain.charAt(0) === "." && domain.substring(1) === hostname.replace(/^[^.]+\./, "");
    };
    selected = function() {
      var cookie, domain, in_domain, in_path, matching, name, path;
      matching = [];
      for (domain in cookies) {
        in_domain = cookies[domain];
        if (!domainMatch(domain, hostname)) {
          continue;
        }
        for (path in in_domain) {
          in_path = in_domain[path];
          if (pathname.indexOf(path) !== 0) {
            continue;
          }
          for (name in in_path) {
            cookie = in_path[name];
            if (typeof cookie.expires === "number" && cookie.expires <= browser.clock) {
              delete in_path[name];
            } else {
              matching.push([domain, path, name, cookie]);
            }
          }
        }
      }
      return matching.sort(function(a, b) {
        return a[1].length - b[1].length;
      });
    };
    this.get = function(name) {
      var match, _i, _len, _ref;
      _ref = selected();
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        match = _ref[_i];
        if (match[2] === name) {
          return match[3].value;
        }
      }
    };
    this.set = function(name, value, options) {
      var in_domain, in_path, maxage, path_without_resource, state, _name, _name2;
      if (options == null) {
        options = {};
      }
      if (options.domain && !domainMatch(options.domain, hostname)) {
        return;
      }
      name = name;
      state = {
        value: value.toString()
      };
      if (options.expires) {
        state.expires = options.expires.getTime();
      } else {
        maxage = options["max-age"];
        if (typeof maxage === "number") {
          state.expires = browser.clock + maxage;
        }
      }
      if (options.secure) {
        state.secure = true;
      }
      if (typeof state.expires === "number" && state.expires <= browser.clock) {
        return this.remove(name, options);
      } else {
        path_without_resource = pathname.match(/.*\//);
        in_domain = cookies[_name = options.domain || hostname] || (cookies[_name] = {});
        in_path = in_domain[_name2 = options.path || path_without_resource] || (in_domain[_name2] = {});
        return in_path[name] = state;
      }
    };
    this.remove = function(name, options) {
      var in_domain, in_path;
      if (options == null) {
        options = {};
      }
      if (in_domain = cookies[options.domain || hostname]) {
        if (in_path = in_domain[options.path || pathname]) {
          return delete in_path[name];
        }
      }
    };
    this.clear = function(options) {
      var in_domain;
      if (options == null) {
        options = {};
      }
      if (in_domain = cookies[hostname]) {
        return delete in_domain[pathname];
      }
    };
    this.update = function(serialized) {
      var cookie, _i, _len, _ref, _results;
      if (!serialized) {
        return;
      }
      if (serialized.constructor === Array) {
        serialized = serialized.join(",");
      }
      _ref = serialized.split(/,(?=[^;,]*=)|,$/);
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        cookie = _ref[_i];
        cookie = deserialize(cookie);
        _results.push(this.set(cookie.name, cookie.value, cookie));
      }
      return _results;
    };
    this.addHeader = function(headers) {
      var header, match;
      header = ((function() {
        var _i, _len, _ref, _results;
        _ref = selected();
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          match = _ref[_i];
          _results.push("" + match[2] + "=" + match[3].value);
        }
        return _results;
      })()).join("; ");
      if (header.length > 0) {
        return headers.cookie = header;
      }
    };
    this.__defineGetter__("pairs", function() {
      var match;
      return ((function() {
        var _i, _len, _ref, _results;
        _ref = selected();
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          match = _ref[_i];
          _results.push("" + match[2] + "=" + match[3].value);
        }
        return _results;
      })()).join("; ");
    });
    this.dump = function(separator) {
      var match;
      if (separator == null) {
        separator = "\n";
      }
      return ((function() {
        var _i, _len, _ref, _results;
        _ref = selected();
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          match = _ref[_i];
          _results.push(serialize(browser, match[0], match[1], match[2], match[3]));
        }
        return _results;
      })()).join(separator);
    };
  }
  return Cookies;
})();
core.HTMLDocument.prototype.__defineGetter__("cookie", function() {
  return this.parentWindow.cookies.pairs;
});
core.HTMLDocument.prototype.__defineSetter__("cookie", function(cookie) {
  return this.parentWindow.cookies.update(cookie);
});
exports.use = function(browser) {
  var access, cookies, dump, extend, load, save;
  cookies = {};
  access = function(hostname, pathname) {
    return new Cookies(browser, cookies, hostname, pathname);
  };
  extend = function(window) {
    return window.__defineGetter__("cookies", function() {
      return access(this.location.hostname, this.location.pathname);
    });
  };
  dump = function() {
    var cookie, domain, in_domain, in_path, name, path, serialized;
    serialized = [];
    for (domain in cookies) {
      in_domain = cookies[domain];
      for (path in in_domain) {
        in_path = in_domain[path];
        for (name in in_path) {
          cookie = in_path[name];
          serialized.push(serialize(browser, domain, path, name, cookie));
        }
      }
    }
    return serialized;
  };
  save = function() {
    var cookie, domain, in_domain, in_path, name, path, serialized;
    serialized = ["# Saved on " + (new Date().toISOString())];
    for (domain in cookies) {
      in_domain = cookies[domain];
      for (path in in_domain) {
        in_path = in_domain[path];
        for (name in in_path) {
          cookie = in_path[name];
          serialized.push(serialize(browser, domain, path, name, cookie));
        }
      }
    }
    return serialized.join("\n");
  };
  load = function(serialized) {
    var cookie, _i, _len, _ref, _results;
    _ref = serialized.split(/\n+/);
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      cookie = _ref[_i];
      if (cookie[0] === "#") {
        continue;
      }
      cookie = deserialize(cookie);
      _results.push(new Cookies(browser, cookies, cookie.domain, cookie.path).set(cookie.name, cookie.value, cookie));
    }
    return _results;
  };
  return {
    access: access,
    extend: extend,
    save: save,
    load: load,
    dump: dump
  };
};
